### PR TITLE
Avoid promoting over-sized intrinsic

### PR DESCRIPTION
**Pull request by cshung**
This PR is meant to fix an assert in `src\jit\morph.cpp:16898` when `JIT\HardwareIntrinsics\X86\Avx\Avx_r\CPAOT-ret.out\Avx_r.dll` is compiled under `crossgen2`

Here is my understanding on what is going on:

When we attempt to promote a struct, we obtain information about the struct though `Compiler::StructPromotionHelper::GetFieldInfo`, without the size check, that code believes `Vector256<Double>` is `TYP_SIMD32` but it actually cannot be because we turned off `Avx`.

Then `Compiler::StructPromotionHelper::TryPromoteStructField` confused, it thinks `Vector256<Double>` is not a `TYP_STRUCT` and let the promotion happen.

Downstream, we failed, `Compiler::fgMorphStructField` discovered that we are promoting a struct field which is also a struct.

My confidence in this code change is just around 30%. It is very possible that something in `crossgen2` is wrong instead.

@dotnet/crossgen-contrib

### CODE REVIEW

**Comment by janvorli** [comment]
cc: @dotnet/jit-contrib

**Comment by tannergooding** [comment]
@cshung, could you share which method in Avx_r is causing the assert? I would suspect its one of the `test.RunUnsupportedScenario` scenarios since that is explicitly calling an unsupported method and checking for `PNSE`.

**Comment by cshung** [comment]
This is the function that when compiling causing the assertion (note that some functions are inlined):

https://github.com/dotnet/coreclr/blob/69a57026b4bcacd8520fffb84842b31ac337e59f/tests/src/JIT/HardwareIntrinsics/X86/Avx/InsertVector128.Double.1.cs#L99

Here is the field that got interpreted as TYP_SIMD32:

https://github.com/dotnet/coreclr/blob/69a57026b4bcacd8520fffb84842b31ac337e59f/tests/src/JIT/HardwareIntrinsics/X86/Avx/InsertVector128.Double.1.cs#L96

**Comment by CarolEidt** [comment]
This will presumably change once we've settled the question of how to handle unsupported types - https://github.com/dotnet/coreclr/pull/23899#issuecomment-551881601 - but I think this is a reasonable fix. To summarize, this fix would not prevent `getBaseTypeAndSizeOfSIMDType` from returning `TYP_SIMD32`, but would prevent us from promoting a field of that type, since it would not fit into a register.

File: src/jit/lclvars.cpp [hunk 1]
```diff
@@ -1978,13 +1978,16 @@ Compiler::lvaStructFieldInfo Compiler::StructPromotionHelper::GetFieldInfo(CORIN
     {
         unsigned  simdSize;
         var_types simdBaseType = compiler->getBaseTypeAndSizeOfSIMDType(fieldInfo.fldTypeHnd, &simdSize);
-        if (simdBaseType != TYP_UNKNOWN)
+        if ((simdSize >= compiler->minSIMDStructBytes()) && (simdSize <= compiler->maxSIMDStructBytes()))
```
Comment (CarolEidt) [inline src/jit/lclvars.cpp]: This should have a comment indicating that we will only promote fields of SIMD types that fit into a SIMD register.

**Comment by tannergooding** [comment]
> the question of how to handle unsupported types 

I don't think that will change this here. That discussion (https://github.com/dotnet/coreclr/pull/23899#issuecomment-551881601) was meant to be around interop handling (where both sides might have an inconsistent view of the world) not around the managed side (where we can always be consistent).

**Comment by CarolEidt** [comment]
@tannergooding - my point was that, if we decide to outlaw unsupported types, then it would presumably be best to not even return `TYP_SIMD32` in this case (i.e. change `getBaseTypeAndSizeOfSIMDType()`) . If, however, we want to support interop for some set of unsupported types (i.e. those for which we can still support ABI requirements even if we don't support intrinsic methods on them), then we *do* need to recognize them.

**Comment by cshung** [comment]
I am a little confused here. Is the change ready to go after adding the comment or we need further discussion? If this needs more discussion then I can try to arrange a [hopefully short] meeting.

**Comment by CarolEidt** [comment]
Sorry @cshung - I guess I wasn't clear. I believe this is the right fix unless and until we decide to change how we handle unsupported types. It looks like it needs some formatting fixes but otherwise LGTM

**Comment by cshung** [comment]
@CarolEidt No worries, it was just me being uncertain about my fix. I will get the comment written and the formatting fixed.

**Pull request merged**